### PR TITLE
[Model Compression] attention mode

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
Add a new sparsity allocator. Add mode `attention` to `L1NormPruner`, `L2NormPruner`, `TaylorFOWeightPruner`.

This allocator is for pruning Q K V projection. Q K V need to be implemented as `torch.nn.Linear` layers, and their weights should have the same size. Then this allocator will sum their weight metric and apply the same masks to these three layers in one attention head.

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py [hunk 1]
```diff
@@ -176,3 +177,108 @@ def _group_metric_calculate(self, group_metrics: Union[Dict[str, Tensor], List[T
...
+
+    def _get_dependency(self):
+        """
+        Populate self.attention_name_groups by running inference on the module graph.
+        Currently, the group inferred AttentionWeightDependency is limited to a set of four weights, with the first
+        three corresponding to :math:`Q_{proj}`, :math:`K_{proj}`, :math:`V_{proj}` (in any order) and the last one being :math:`{Output}_{proj}`.
+        """
+        try:
+            graph = self.pruner.generate_graph(dummy_input=self.dummy_input)
+            dependency_tracer = AttentionWeightDependency(traced_model=graph.trace)
```
Comment (zheng-ningxin) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:201]: The logic of `AttentionWeightDependency` is not robust enough, maybe we can implement a function to find the exact same subgraph with `SelfAttention`?

Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:201]: you are right, I will check if `SelfAttention` implementation is similar in other transformers, and find a better way to match the attention.

Comment (linbinskn) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:201]: What is the constraint of `AttentionWeightDependency`, only support specific module type (such as only torch.nn.Module, not support F function) or also not support other transformer implementation (such as only support Huggingface)? May be we should provide more notes to help user notice these constraints.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py [hunk 2]
```diff
@@ -176,3 +177,108 @@ def _group_metric_calculate(self, group_metrics: Union[Dict[str, Tensor], List[T
...
+        # Group the QKV metrics in one head.
+        for idx, names in enumerate(self.attention_name_groups):
+            group_metric_dict = OrderedDict()
+            for name in names[0: 3]:
+                # If the Linear will be pruned, add it to this group.
+                if name in metrics:
+                    group_metric_dict[name] = metrics[name]
+                    if self.continuous_mask:
+                        group_metric_dict[name] *= self._compress_mask(self.pruner.get_modules_wrapper()[name].weight_mask)
+            grouped_metrics[idx] = group_metric_dict
```
Comment (linbinskn) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:233]: It seems that we don't support `op_names` in attention mode. If module is pruned by setting `op_names` in config_list while `self.attention_name_groups` recording all name of Q, K, V, error may be raised on line 237 since some value of `grouped_metrics` will be empty.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py [hunk 3]
```diff
@@ -176,3 +177,108 @@ def _group_metric_calculate(self, group_metrics: Union[Dict[str, Tensor], List[T
...
+        # Calculate group metric.
+        for idx, group_metric_dict in grouped_metrics.items():
+            group_metric = self._group_metric_calculate(group_metric_dict)
+            for name in group_metric_dict.keys():
+                metrics[name] = group_metric
+
+            # Special rule based output layer mask generation.
+            output_layer_name = self.attention_name_groups[idx][-1]
+            if output_layer_name in metrics and self.dim and len(self.dim) == 1 and self.dim[0] == 0:
+                # Don't use colculated metric, use group metric.
```
Comment (linbinskn) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:244]: colculated->calculated

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 4]
```diff
@@ -635,7 +653,7 @@ def trainer(model: Module, optimizer: Optimizer, criterion: Callable[[Tensor, Te
     training_batches : int
         The batch number used to collect activations.
     mode : str
-        'normal', 'dependency_aware' or 'global'.
+        'normal', 'dependency_aware', 'global', 'attention'.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:656]: is 'attention' qualified to be one mode?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py (hunk 1 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:201]: For checking several Huggingface transformer implementations, `bert`, `roberta`, `gpt_neo`, `bart` `bait`...have similar `Q K V Projection Structure`.
And implementations from other repos or even in Huggingface may not have the similar structure, so I think this mode can only support huggingface implementations right now.
I will add more information to explain these constraints.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:233]: you are right, this place will raise an error, I will fix it.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 4 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:656]: I think it is a pruning mode because it will prune multiple layers as a whole
